engine/src/types.rs

```diff
@@ -986,6 +986,8 @@ enum Layer {
     Map,
 }
 
+/// A types for field values that can be nested within another type such as an
+/// array or map.
```

Comment:
A compound type can actually describe all types, its just a "flattened" representation to avoid having to box inner types for maps and arrays. Without the box it would make an infinite size type